### PR TITLE
Added geneve header options

### PR DESCRIPTION
**Pull request by HareshKhandelwal**
currently, we dont have geneve options working.
This patch aims to fix them.

Unit test cases:

With Options:
```
sendp(Ether(dst='0c:42:a1:d1:da:98',src='40:a6:b7:0b:e9:b1')/Dot1Q(vlan=304,type='IPv4')/IP(proto=17,src='172.17.2.161',dst='172.17.2.32')/UDP(sport=57025,dport=6081)/GENEVE(proto=0x6558,vni=4,options=GeneveOptions(classid=0x0102,type=0x80,data=b'\x00\x01\x00\x02'))/Ether(dst='fa:16:3e:81:fa:0f',src='fa:16:3e:e7:a9:8a',type=0x0800)/IP(proto=1,dst='192.168.2.181',src='192.168.2.92')/ICMP(type=8),iface="wlp0s20f3")


16:19:54.952429 40:a6:b7:0b:e9:b1 > 0c:42:a1:d1:da:98, ethertype 802.1Q (0x8100), length 104: vlan 304, p 0, ethertype IPv4, (tos 0x0, ttl 64, id 1, offset 0, flags [none], proto UDP (17), length 86)
    172.17.2.161.57025 > 172.17.2.32.6081: [udp sum ok] Geneve, Flags [none], vni 0x4, proto TEB (0x6558), options [class Open Virtual Networking (OVN) (0x102) type 0x80(C) len 8 data 00010002]
	fa:16:3e:e7:a9:8a > fa:16:3e:81:fa:0f, ethertype IPv4 (0x0800), length 42: (tos 0x0, ttl 64, id 1, offset 0, flags [none], proto ICMP (1), length 28)
    192.168.2.92 > 192.168.2.181: ICMP echo request, id 0, seq 0, length 8
```

Without options:
```
sendp(Ether(dst='0c:42:a1:d1:da:98',src='40:a6:b7:0b:e9:b1')/Dot1Q(vlan=304,type='IPv4')/IP(proto=17,src='172.17.2.161',dst='172.17.2.32')/UDP(sport=57025,dport=6081)/GENEVE(proto=0x6558,vni=4)/Ether(dst='fa:16:3e:81:fa:0f',src='fa:16:3e:e7:a9:8a',type=0x0800)/IP(proto=1,dst='192.168.2.181',src='192.168.2.92')/ICMP(type=8),iface="wlp0s20f3")

16:21:08.305638 40:a6:b7:0b:e9:b1 > 0c:42:a1:d1:da:98, ethertype 802.1Q (0x8100), length 96: vlan 304, p 0, ethertype IPv4, (tos 0x0, ttl 64, id 1, offset 0, flags [none], proto UDP (17), length 78)
    172.17.2.161.57025 > 172.17.2.32.6081: [udp sum ok] Geneve, Flags [none], vni 0x4, proto TEB (0x6558)
	fa:16:3e:e7:a9:8a > fa:16:3e:81:fa:0f, ethertype IPv4 (0x0800), length 42: (tos 0x0, ttl 64, id 1, offset 0, flags [none], proto ICMP (1), length 28)
    192.168.2.92 > 192.168.2.181: ICMP echo request, id 0, seq 0, length 8

>>> packet = GeneveOptions(classid=0x0102,type=0x80,data=b'\x00\x01\x00\x02')
>>> packet.show()
###[ Geneve Options ]### 
  classid   = Open Virtual Networking (OVN)
  type      = 0x80
  reserved  = 0
  length    = None
  data      = '\x00\x01\x00\x02'

>>> packet.show2()
###[ Geneve Options ]### 
  classid   = Open Virtual Networking (OVN)
  type      = 0x80
  reserved  = 0
  length    = 1
  data      = '\x00\x01\x00\x02'

>>> 

>>> packet = GENEVE(proto=0x6558,vni=4,options=GeneveOptions(classid=0x0102,type=0x80,data=b'\x00\x04\x00\x05'))
>>> packet.show()
###[ GENEVE ]### 
  version   = 0
  optionlen = None
  oam       = 0
  critical  = 0
  reserved  = 0
  proto     = TEB
  vni       = 0x4
  reserved2 = 0x0
  \options   \
   |###[ Geneve Options ]### 
   |  classid   = Open Virtual Networking (OVN)
   |  type      = 0x80
   |  reserved  = 0
   |  length    = None
   |  data      = '\x00\x04\x00\x05'

>>> packet.show2()
###[ GENEVE ]### 
  version   = 0
  optionlen = 2
  oam       = 0
  critical  = 0
  reserved  = 0
  proto     = TEB
  vni       = 0x4
  reserved2 = 0x0
  \options   \
   |###[ Geneve Options ]### 
   |  classid   = Open Virtual Networking (OVN)
   |  type      = 0x80
   |  reserved  = 0
   |  length    = 1
   |  data      = '\x00\x04\x00\x05'

>>> 
```

**Checklist:**

-   [x] If you are new to Scapy: I have checked [CONTRIBUTING.md](https://github.com/secdev/scapy/blob/master/CONTRIBUTING.md) (esp. section submitting-pull-requests)
-   [x] I squashed commits belonging together
-   [x] I added unit tests or explained why they are not relevant
-   [ ] I executed the regression tests for Python2 and Python3 (using `tox` or, `cd test && ./run_tests_py2, cd test && ./run_tests_py3`)
-   [ ] If the PR is still not finished, please create a [Draft Pull Request](https://github.blog/2019-02-14-introducing-draft-pull-requests/)

fixes #xxx

### CODE REVIEW

**Comment by HareshKhandelwal** [comment]
Hi @guedou , Mistakenly pull request#3287 got closed. Raised this new one. 
Addressed your comments on re add author name, named options variable as you suggested. Provided unit test cases. 
Please have a review.

Thanks

**Review by guedou — COMMENTED** [review]
This looks good to me. 

Would you mind adding some simple unit tests such as https://github.com/secdev/scapy/blob/master/test/contrib/geneve.uts#L8 ?

**Comment by HareshKhandelwal** [comment]
Hi @guedou 

I am writing the unit test cases and got stuck at one place. Can you please provide your opinion where i am going wrong!

Unit test case is 

s = Ether(dst='0c:42:a1:d1:da:98',src='40:a6:b7:0b:e9:b1')/Dot1Q(vlan=304,type='IPv4')/IP(proto=17,src='172.17.2.161',dst='172.17.2.32')/UDP(sport=57025,dport=6081)/GENEVE()/Ether(dst='fa:16:3e:81:fa:0f',src='fa:16:3e:e7:a9:8a',type=0x0800)/IP(proto=1,dst='192.168.2.181',src='192.168.2.92')/ICMP(type=8)

While i am able to send proper packet with above packet construct, s.show2() doesnt show any other Encap post Geneve. 

>>> s.show2()
###[ Ethernet ]### 
  dst       = 0c:42:a1:d1:da:98
  src       = 40:a6:b7:0b:e9:b1
  type      = n_802_1Q
###[ 802.1Q ]### 
     prio      = 0
     id        = 0
     vlan      = 304
     type      = IPv4
###[ IP ]### 
        version   = 4
        ihl       = 5
        tos       = 0x0
        len       = 78
        id        = 1
        flags     = 
        frag      = 0
        ttl       = 64
        proto     = udp
        chksum    = 0x1dbb
        src       = 172.17.2.161
        dst       = 172.17.2.32
        \options   \
###[ UDP ]### 
           sport     = 57025
           dport     = geneve
           len       = 58
           chksum    = 0x298a
###[ GENEVE ]### 
              version   = 0
              optionlen = 0
              oam       = 0
              critical  = 0
              reserved  = 0
              proto     = TEB
              vni       = 0x0
              reserved2 = 0x0
              \options   \
               |###[ Geneve Options ]### 
               |  classid   = 0xfa16
               |  type      = 0x3e
               |  reserved  = 4
               |  length    = 1
               |  data      = '\\xfa\x0f\\xfa\x16>穊\x08\x00E\x00\x00\x1c\x00\x01\x00\x00@\x01\\xf4~\\xc0\\xa8\x02\\\\xc0\\xa8\x02\\xb5\x08\x00\\xf7\\xff\x00\x00\x00\x00'

>>> 

This makes existing unit test case fail (Which check further encap of Ether). I have checked packet length but didn't find any issue, may be missing anything. Can you please help to identify the issue?

Thanks

**Comment by guedou** [comment]
It seens that GeneveOptions is consuming all the remaining data. You likely need to define `extract_padding` as done in https://github.com/secdev/scapy/blob/adaa923db82be88e5bd84b4046faa23f2b69d0ed/scapy/layers/dns.py#L524

**Comment by HareshKhandelwal** [comment]
Hi @guedou 

I did rework on clean up and updated unit cases as well. All unit cases passing now.

Thanks
-Haresh

File: scapy/contrib/geneve.py [hunk 1]
```diff
@@ -54,17 +72,16 @@ class GENEVE(Packet):
...
-            optionlen = (len(self.options) + 3) // 4
-            p = chb(optionlen & 0x2f | orb(p[0]) & 0xc0) + p[1:]
-        return p
+        if self.optionlen is None:
+            tmp_len = (len(p) - 8) // 4
+            p = chb(tmp_len & 0x2f | orb(p[0]) & 0xc0) + p[1:]
+        return p + pay
 
     def answers(self, other):
+        print("in geneve answer")
```
Comment (gpotter2) [inline scapy/contrib/geneve.py:84]: Forgotten debug print

Comment (HareshKhandelwal) [reply scapy/contrib/geneve.py]: my bad..removed now

**Comment by HareshKhandelwal** [comment]
@gpotter2 @guedou can you please help with your reviews? Thanks

**Comment by polydroi** [comment]
LGTM

**Review by gpotter2 — APPROVED** [review]

**Review by guedou — APPROVED** [review]

**Pull request merged**